master/internal/db/postgres_trial_metrics.go

```diff
@@ -81,26 +138,97 @@ WHERE trial_id = $1
 	return int(affectedRows), nil
 }
 
-func (db *PgDB) addRawMetrics(ctx context.Context, tx *sqlx.Tx, metricsBody *map[string]interface{},
+// addMetricsWithMerge inserts a set of metrics to the database and returns the metric id,
+// and the results.
+func (db *PgDB) addMetricsWithMerge(ctx context.Context, tx *sqlx.Tx, mBody *metricsBody,
+	runID, trialID, lastProcessedBatch int32, mType model.MetricType,
+) (metricID int, addedMetrics *metricsBody, err error) {
+	var existingBodyJSON model.JSONObj
+	err = tx.QueryRowContext(ctx, `
+SELECT COALESCE((SELECT metrics FROM metrics
+WHERE archived = false
+AND total_batches = $1
+AND trial_id = $2
+AND partition_type = $3
+AND custom_type = $4), NULL)`,
```

Comment:
should this be `SELECT ... FOR UPDATE`?

Reply:
yeah, we do need to lock these rows in case of updates.